client-js/src/client/errors.js

```diff
@@ -0,0 +1,187 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import 'spine-web-client-proto/spine/base/error_pb';
+// noinspection JSUnresolvedVariable
+const SpineBaseError = proto.spine.base.Error;
+
+/**
+ * The base `spine-web-client` error type. This error type is only used directly when a
+ * more appropriate category is not defined for the offending error.
+ *
+ * @extends Error
+*/
+export class SpineError extends Error {
+
+  /**
+   * @param {!string} message the human-readable error message
+   * @param {*=} cause        the reason why this error occurred
+   */
+  constructor(message, cause) {
+    super(message);
+    this.name = this.constructor.name;
+    this._cause = cause;
+  }
+
+  /**
+   * @return {*|undefined} The cause of this error, if available.
+   */
+  getCause() {
+    return this._cause;
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint fails due to the
+ * connection problems.
+ *
+ * Can be caused by an incorrect server address, lack of network connectivity or
+ * if the response is not received from the server.
+ *
+ * @extends SpineError
+ */
+export class ConnectionError extends SpineError {
+
+  /**
+   * @param {!Error} error the error caught from {@code fetch} invocation
+   */
+  constructor(error) {
+    super(error.message, error);
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results
+ * with a response with `5xx` status code.
+ *
+ * @extends SpineError
+ */
+export class ServerError extends SpineError {
+
+  /**
+   * @param {!Response} response the server response caused this error
+   */
+  constructor(response) {
+    super(response.statusText, response);
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results
+ * with a response with `4xx` status code.
+ *
+ * @extends SpineError
+ */
+export class ClientError extends SpineError {
+
+  /**
+   * @param {!string} message the human-readable error message
+   * @param {*=} cause        the reason why this error occurred
+   */
+  constructor(message, cause) {
+    super(message, cause);
+  }
+}
+
+/**
+ * An error which occurs when sending off a command to Spine server endpoint.
+ *
+ * @extends SpineError
+ */
+export class CommandHandlingError extends SpineError {
+
+  /**
+   * @param {!string} message the human-readable error message
+   * @param {!Error|SpineError|SpineBaseError} error the reason why this error occurred
+   */
+  constructor(message, error) {
+    super(message, error);
+  }
+
+  /**
+   * Returns `true` if the command wasn't accepted by the server; returns `false`
+   * if this is not guaranteed.
+   *
+   * A command is assumed neglected if this error is caused by the `ClientError`
+   * or the `SpineBaseError`.
+   *
+   * @return {boolean}
+   */
+  assureCommandNeglected() {
```

Comment:
Can there be a better name for this method?

Reply:
This method returns `true` only when it is guaranteed that the command was ignored. When it returns `false`, the command can still be ignored. What do you think about `isCommandDefinitelyIgnored()`?